corehq/apps/app_manager/dbaccessors.py

```diff
@@ -176,6 +176,35 @@ def get_built_app_ids_for_app_id(domain, app_id, version=None):
     return [result['id'] for result in results]
 
 
+def get_latest_built_app_ids_and_versions(domain, app_id=None):
+    """
+    Returns all the latest build app_ids and version in a dictionary:
+    {app_id: latest_version}
+    """
+    from .models import Application
+    key = [domain, app_id] if app_id else [domain]
+
+    results = Application.get_db().view(
+        'app_manager/saved_app',
+        startkey=key + [{}],
+        endkey=key,
+        descending=True,
+        reduce=False,
+        include_docs=False,
+    ).all()
+
+    latest_ids_and_versions = {}
+    for result in results:
+        app_id = result['key'][1]
+        version = result['key'][2]
+
+        # Since we have sorted, we know the first instance is the latest version
+        if app_id not in latest_ids_and_versions:
```

Comment:
If its the same as the one passed in then why do we need to loop?

On 29 January 2016 at 18:10, Ben <notifications@github.com> wrote:

> In corehq/apps/app_manager/dbaccessors.py
> <https://github.com/dimagi/commcare-hq/pull/10107#discussion_r51278654>:
>
> > +    results = Application.get_db().view(
> > +        'app_manager/saved_app',
> > +        startkey=key + [{}],
> > +        endkey=key,
> > +        descending=True,
> > +        reduce=False,
> > +        include_docs=False,
> > +    ).all()
> > +
> > +    latest_ids_and_versions = {}
> > +    for result in results:
> > +        app_id = result['key'][1]
> > +        version = result['key'][2]
> > +
> > +        # Since we have sorted, we know the first instance is the latest version
> > +        if app_id not in latest_ids_and_versions:
>
> actually i'm not sure what to change it to. the app_id is indeed the same
> value as the one passed in. If there's not one passed in, then it's all the
> app ids in the domain. i'll add a comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10107/files#r51278654>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa